server.go

```diff
@@ -107,7 +107,16 @@ func (s *Server) recv(c *net.UDPConn) {
 		return
 	}
 	buf := make([]byte, 65536)
-	for !s.shutdown {
+	var shutdown bool
+	s.shutdownLock.RLock()
+	defer s.shutdownLock.RUnlock()
+	for {
+		s.shutdownLock.RLock()
```

Comment:
Won't you end up with a double RLock here? Why lock here instead of just in the for loop?